compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+            }
+            Artifact::PilWithConstants(_) => {
+                return Err(vec![
+                    "Witness generation requires arguments, call generate_witness() instead!"
+                        .into(),
+                ])
+            }
+            Artifact::GeneratedWitness(_) => {
+                return Err(vec![
+                    "Proof generation requires arguments, call prove() instead!".into(),
+                ])
+            }
+            Artifact::Proof(_) => panic!("Last pipeline step!"),
+        });
+        Ok(())
+    }
+
+    fn maybe_write_pil<C: Display>(&self, content: &C, suffix: &str) -> Result<(), Vec<String>> {
+        if let Some(output_dir) = &self.output_dir {
+            let name = self
+                .name
+                .as_ref()
+                .expect("name must be set if output_dir is set");
+            let output_file = output_dir.join(format!("{name}{suffix}.pil"));
+            if !output_file.exists() || self.force_overwrite {
+                fs::write(&output_file, format!("{content}")).map_err(|e| {
+                    vec![format!(
+                        "Error writing {}: {e}",
+                        output_file.to_str().unwrap()
+                    )]
+                })?;
+                self.log(&format!("Wrote {}.", output_file.to_str().unwrap()));
+            } else {
+                return Err(vec![format!(
+                    "{} already exists! Use --force to overwrite.",
+                    output_file.to_str().unwrap()
+                )]);
+            }
+        }
+        Ok(())
+    }
+
+    fn stage(&self) -> Stage {
+        match self.artifact.as_ref().unwrap() {
+            Artifact::AsmFile(_) => Stage::AsmFile,
+            Artifact::AsmString(_, _) => Stage::AsmString,
+            Artifact::Parsed(_, _) => Stage::Parsed,
+            Artifact::Resolved(_) => Stage::Resolved,
+            Artifact::AnalyzedAsm(_) => Stage::AnalyzedAsm,
+            Artifact::PilConstraints(_) => Stage::PilConstraints,
+            Artifact::Graph(_) => Stage::Graph,
+            Artifact::Linked(_) => Stage::Linked,
+            Artifact::PilFile(_) => Stage::PilFile,
+            Artifact::PilString(_) => Stage::PilString,
+            Artifact::AnaylyzedPil(_) => Stage::AnalyzedPil,
+            Artifact::OptimzedPil(_) => Stage::OptimizedPil,
+            Artifact::PilWithConstants(_) => Stage::PilWithConstants,
+            Artifact::GeneratedWitness(_) => Stage::GeneratedWitness,
+            Artifact::Proof(_) => Stage::Proof,
+        }
+    }
+
+    fn advance_to(&mut self, target_stage: Stage) -> Result<(), Vec<String>> {
+        while self.stage() != target_stage {
+            self.advance()?;
+        }
+        Ok(())
+    }
+
+    pub fn generate_witness<Q: QueryCallback<T>>(
+        &mut self,
+        query_callback: Q,
+        external_witness_values: Vec<(&str, Vec<T>)>,
+    ) -> Result<(), Vec<String>> {
```

Comment:
Another way to deal with arguments would be to have an API like this:
```rust
let proof = Pipeline::default()
    .with_inputs(inputs)
    .with_backend(backend)
    .proof();
```

Basically, for all arguments, you'd have to provide them before the step is run, and running it fails if required arguments are missing. The advantage would be that arguments can be optional (e.g. `external_witness_values`) or there can be multiple ways to specify the same argument (e.g. `with_inputs()` and `with_query_callback()`).

Think that would be even better, but maybe it's fine to do that in a follow-up PR!

Reply:
Implemented this suggestion in #834. We could merge it into this branch or keep as a follow-up PR.